### DISCUSSION TITLE
useEffect order of execution in _app.tsx

### DISCUSSION THREAD

**Discussion by json9512** [vercel/next.js#40078] [Help] [answered]
Hi,

I have a question about the order of execution in `_app.tsx`.

According to this [post](https://blog.logrocket.com/post-hooks-guide-react-call-order/), react executes the `useEffect`s from the child component up to the parent component.

However, when I run `useEffect` from `_app.tsx`, `PageComponent`, `SomeComponentInPageComponent`, the order of execution is `_app.tsx` -> `SomeComponentInPageComponent` -> `PageComponent`.

(What I expected was `SomeComponentInPageComponent` -> `PageComponent` -> `_app.tsx`)

It seems like `_app.tsx` is an exception to the order of execution rule, and I would like to know the reason (if there is one) or the cause of such behavior. (It's just out of curiosity)

Thanks,
Son

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Which versions are you using? I cannot reproduce this:

<img width="546" alt="Screenshot 2022-08-30 at 10 00 25" src="https://user-images.githubusercontent.com/21013447/187383045-c2b185a4-243b-4e18-be6d-92e87ffb109b.png">

There's no special behavior going on, but perhaps there's more to your app than just `useEffect` calls?

```tsx
// _app
function MyApp({ Component, pageProps }: AppProps) {
  useEffect(() => {
    console.log("_app");
  }, []);
  return <Component {...pageProps} />;
}

// at pages index
const Child = () => {
  useEffect(() => {
    console.log("_child");
  }, []);

  return <div>child</div>;
};

const Home: NextPage = () => {
  useEffect(() => {
    console.log("_home");
  }, []);
  return (
    <div>
      Home <Link href="/blog">blog</Link>
      <div>more stuff</div>
      <Child />
    </div>
  );
};
```

There's React Strict Mode activated, which duplicates the logs, but that's about it. 

Strictly speaking, console logging is not necessary a good method to analyse order of execution, but as you can see, it is good enough here, how are you measuring it?

**Reply by json9512** [reply]
@icyJoseph hi, I think I found the logic issue with my `_app.tsx`. I think it's because I initialized the page component prior to returning it in the `_app.tsx`.

currently using:
- "next": "^11.1.4",
- "react": "^17.0.2",

The structure of the `_app.tsx` for my project is:
```typescript
// _app.tsx
function MyApp() {

  useEffect(() => {
    console.log('myApp')
  }, []);


  // ... some logic
  // the initialization here changes the order of execution
  // If I return the page component here, the order of execution is not changed.
  let app = (
    <SomeBigComponent>
      <InnerComponent /> // <- this is the typical Page Component
    </SomeBigComponent>
  );
 
  // ... some other logic that adds ErrorBoundary and PersistGate from 'redux-persist/integration/react'
  return app;
}

export default wrapper.withRedux(MyApp); // <- from next-redux-wrapper (cause)

// Page Component
function SampleInnerComponent() {

  useEffect(() => {
    console.log('sample inner component');
  }, []);  

  return (
    <SampleChildComponent />
  );
}

// Child Component
function SampleChildComponent() {

  useEffect(() => {
    console.log('sample child component');
  }, []);  

  return <div>Test</div>
}

// Above will log
'myApp'
'sample child component'
'sample inner component'
```

~~It was a logic flaw after all.~~
It was an issue with the redux wrapper.

Thanks for the help! 👍

**Reply by icyJoseph** [reply]
Ah, as usual a 3rd party non included in the original question hehe, good that you figured it out! Though it still sounds fishy to me.

**Reply by json9512** [reply]
yeah. sorry for the ambiguous question 🙃

I tracked further and figured that it was actually the `PersistGate` section of the logic.

```typescript
// MyApp component
// ...
// ... some other logic that adds ErrorBoundary and PersistGate from 'redux-persist/integration/react'

if (someCondition) {
  app = <PersistGate>{app}</PersistGate>
}

return app
```

I guess this question should be marked as solved since it's got nothing to do with next.js 👍
